internal/inventory/tagprocessor.go

```diff
@@ -0,0 +1,364 @@
+	adjustLastReadOnByOrigin bool
+
+	// debugLogEnabled is used to be able to only log things when Debug logging is enabled
+	// note: this should be something that is able to be determined via the logger.LoggingClient,
+	// however currently EdgeX does not support querying the log level
+	// see: https://github.com/edgexfoundry/go-mod-core-contracts/issues/294
+	debugLogEnabled bool
+}
+
+type TagProcessor struct {
+	lc        logger.LoggingClient
+	inventory map[string]*Tag
+	config    processorConfig
+}
+
+// NewTagProcessor creates a tag processor and pre-loads its mobility profile
+func NewTagProcessor(lc logger.LoggingClient, cfg ConsulConfig, tags []StaticTag) *TagProcessor {
+	tp := &TagProcessor{
+		lc:        lc,
+		inventory: make(map[string]*Tag),
+	}
+	tp.UpdateConfig(cfg)
+
+	for _, t := range tags {
+		tp.inventory[t.EPC] = t.asTagPtr()
+	}
+
+	return tp
+}
+
+func (tp *TagProcessor) UpdateConfig(cfg ConsulConfig) {
+	as := cfg.ApplicationSettings
+	profile := newMobilityProfile(as.MobilityProfileSlope, as.MobilityProfileThreshold, as.MobilityProfileHoldoffMillis)
+	aliases := cfg.Aliases
+	delete(aliases, "")
+
+	logLevel := strings.ToUpper(cfg.Writable.LogLevel)
+	tp.config = processorConfig{
+		adjustLastReadOnByOrigin: as.AdjustLastReadOnByOrigin,
+		departedThresholdSeconds: as.DepartedThresholdSeconds,
+		ageOutHours:              as.AgeOutHours,
+		debugLogEnabled:          logLevel == contract.DebugLog || logLevel == contract.TraceLog,
+		profile:                  profile,
+		aliases:                  aliases,
+	}
+}
+
+// ProcessReport takes an incoming ROAccessReport and processes each TagReportData.
+// For every TagReportData it will update the corresponding tag our in-memory tag database
+// based on the latest information.
+func (tp *TagProcessor) ProcessReport(r *llrp.ROAccessReport, info ReportInfo) (events []Event, snapshot []StaticTag) {
+	if tp.config.adjustLastReadOnByOrigin {
+		// offsetMicros is an adjustment of timestamps
+		// based on when the device service first saw the message
+		// compared to when the sensor said it sent it.
+		// This can be affected by the latency,
+		// but hopefully that value has relatively little jitter.
+		// If a sensor thinks the timestamp is in the future,
+		// this will adjust the times to be standardized
+		// against all other sensors in the system.
+		var lastSeenMicros int64
+		for _, rt := range r.TagReportData {
+			if rt.LastSeenUTC != nil && int64(*rt.LastSeenUTC) > lastSeenMicros {
+				lastSeenMicros = int64(*rt.LastSeenUTC)
+			}
+		}
+		if lastSeenMicros > 0 {
+			// divide originNanos by 1000 to get to micros
+			info.offsetMicros = (info.OriginNanos / 1000) - lastSeenMicros
+		}
+	}
+
+	for _, rt := range r.TagReportData {
+		if event := tp.processData(&rt, info); event != nil {
+			events = append(events, event)
+		}
+	}
+	return events, tp.snapshot()
+}
+
+// ReportInfo holds both pre-existing as well as computed metadata about an incoming ROAccessReport
```

Comment:
This seems out of place and buried in the middle of this large file. Best to extract ReportInfo code into it's own file.